src/index.ts

```diff
@@ -44,3 +45,62 @@ export function getServiceData(apiName: string, beaconName: string, chain: strin
     },
   };
 }
+
+interface PrivateKeySender {
+  privateKey: string;
+}
+
+interface MnemonicSender {
+  mnemonic: string;
+  derivationPath?: string;
+}
+
+type SenderAccount = PrivateKeySender | MnemonicSender;
+
+export function isPrivateKeySender(senderAccount: SenderAccount): senderAccount is PrivateKeySender {
+  return Object.prototype.hasOwnProperty.call(senderAccount, 'privateKey');
+}
+
+interface WhitelistBeaconReaderResult {
+  beaconId: string;
+  beaconReaderAddress: string;
+  expirationTimestamp: number;
+  indefiniteWhitelistStatus: boolean;
+}
+
+export async function whitelistBeaconReader(
+  beaconId: string,
+  beaconReaderAddress: string,
+  chain: string,
+  providerUrl: string,
+  senderAccount: SenderAccount
+): Promise<WhitelistBeaconReaderResult> {
+  const goBeaconWhitelister = goSync(() =>
+    require(`@api3dao/utility-contracts/SelfServeRrpBeaconServerWhitelister/deployments/${chain}/SelfServeRrpBeaconServerWhitelister.json`)
+  );
+  if (!isGoSuccess(goBeaconWhitelister)) throw new Error(`No whitelister contract found for chain: ${chain}`);
+  const beaconWhitelister = goBeaconWhitelister.data;
+
+  const provider = new ethers.providers.JsonRpcProvider(providerUrl);
+  const wallet = isPrivateKeySender(senderAccount)
+    ? new ethers.Wallet(senderAccount.privateKey)
+    : ethers.Wallet.fromMnemonic(senderAccount.mnemonic, senderAccount.derivationPath);
+  const connectedWallet = wallet.connect(provider);
+  const whitelisterContract = new ethers.Contract(beaconWhitelister.address, beaconWhitelister.abi).connect(
+    connectedWallet
+  );
+
+  await whitelisterContract.whitelistReader(beaconId, beaconReaderAddress);
+  const filter = whitelisterContract.filters.WhitelistedReader(beaconId, beaconReaderAddress);
+  // TODO: Simplify
+  const { args } = await new Promise((resolve) => provider.once(filter, resolve)).then((rawLog) =>
+    whitelisterContract.interface.parseLog(rawLog as ethers.Event)
+  );
```

Comment:
I've seen this approach on other places and I have always wondered if we should be subscribing to the event before actually triggering instead of after like in here 🤔

Reply:
I am not completely sure about it too :D But it seems to work reliably...

Reply:
I guess the thing is that there is always enough time between when the transaction is sent to the blo0ckchain provider (line 94) and when the trasansaction gets mined (line 96) . 

SO having it like this is OK to me.